libfreerdp/core/gateway/arm.c

```diff
@@ -0,0 +1,348 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * Azure Virtual Desktop Gateway / Azure Resource Manager
+ *
+ * Copyright 2023 Michael Saxl <mike@mwsys.mine.bz>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <freerdp/config.h>
+
+#include <winpr/assert.h>
+
+#include <winpr/crt.h>
+#include <winpr/synch.h>
+#include <winpr/print.h>
+#include <winpr/stream.h>
+#include <winpr/winsock.h>
+#include <winpr/cred.h>
+
+#include <freerdp/log.h>
+#include <freerdp/error.h>
+#include <freerdp/utils/ringbuffer.h>
+#include <freerdp/utils/smartcardlogon.h>
+
+#include "wst.h"
+#include "websocket.h"
+#include "http.h"
+#include "../credssp_auth.h"
+#include "../proxy.h"
+#include "../rdp.h"
+#include "../../crypto/opensslcompat.h"
+#include "rpc_fault.h"
+#include "../utils.h"
+#include "../settings.h"
+
+#ifdef WITH_CJSON
+#include <cjson/cJSON.h>
+#endif
+
+struct rdp_arm
+{
+	rdpContext* context;
+	rdpSettings* settings;
+	rdpTls* tls;
+	HttpContext* http;
+};
+
+typedef struct rdp_arm rdpArm;
+
+#define TAG FREERDP_TAG("core.gateway.arm")
+
+static BOOL arm_tls_connect(rdpArm* arm, rdpTls* tls, int timeout)
+{
+	WINPR_ASSERT(arm);
+	WINPR_ASSERT(tls);
+	int sockfd = 0;
+	long status = 0;
+	BIO* socketBio = NULL;
+	BIO* bufferedBio = NULL;
+	rdpSettings* settings = arm->settings;
+	const char* peerHostname = arm->settings->GatewayHostname;
```

Comment:
* better `WINPR_ASSERT` all elements to be accessed
* for settings better use `freerdp_settings_get_*` for new code